### PR TITLE
Generate tweets from a content with VercelAI #928

### PR DESCRIPTION
**Pull request by hakanayata**
Post-generation using chatGPT-4
- API Route for streaming response
- Dynamic input fields (number of posts, character limit, and language)
- Response is rendered as textarea

### CODE REVIEW

**Comment by hakanayata** [comment]
Please only take the following files into consideration:
- dashboard/pages/api/route-tweet-gen.ts
- dashboard/pages/[endpoint].tsx

File: apps/dashboard/pages/[endpoint].tsx [hunk 1]
```diff
@@ -236,6 +288,110 @@ const ModelPage: FC<ModelPageProps> = ({ endpoint }) => {
...
+                            key={idx}
+                            value={genPost}
+                            onChange={e => console.log(e.target.value)}
+                          />
+                        )
+                      })
+                    }
+                  </Box>
+                </Box>
+              }
```
Comment (7alip) [inline apps/dashboard/pages/[endpoint].tsx:393]: Could you please move these sections to a separate component, as we did with `PostSentenceForm`? You should probably also move the `useCompletion` hook part and all associated states to the new component. We can keep the `[endpoint]` page clean using this method.

This component may be useful elsewhere in the project, such as in samen.

File: apps/dashboard/pages/[endpoint].tsx [hunk 2]
```diff
@@ -236,6 +288,110 @@ const ModelPage: FC<ModelPageProps> = ({ endpoint }) => {
...
+                  <FormLabel>Content</FormLabel>
+                  <Textarea
+                    name='prompt'
+                    placeholder='Enter a content...'
+                    value={input}
+                    onChange={handleInputChange}
+                    mb={4}
+                    size='xs'
+                    required
+                  />
```
Comment (7alip) [inline apps/dashboard/pages/[endpoint].tsx:304]: You could wrap this with [`FormControl`](https://chakra-ui.com/docs/components/form-control/usage) component

File: apps/dashboard/pages/[endpoint].tsx [hunk 3]
```diff
@@ -236,6 +288,110 @@ const ModelPage: FC<ModelPageProps> = ({ endpoint }) => {
...
+                  <Flex gap={10} mb={4} >
+                    <Box>
```
Comment (7alip) [inline apps/dashboard/pages/[endpoint].tsx:306]: Can be wrapped with [`FormControl (Number input example)`](https://chakra-ui.com/docs/components/form-control/usage#number-input-example)

File: apps/dashboard/pages/[endpoint].tsx [hunk 4]
```diff
@@ -236,6 +288,110 @@ const ModelPage: FC<ModelPageProps> = ({ endpoint }) => {
...
+                        onChange={(a, b) => setCharLimit(b)}
+                      >
+                        <NumberInputField />
+                        <NumberInputStepper>
+                          <NumberIncrementStepper />
+                          <NumberDecrementStepper />
+                        </NumberInputStepper>
+                      </NumberInput>
+                    </Box>
+                    <Box flex='1'>
```
Comment (7alip) [inline apps/dashboard/pages/[endpoint].tsx:338]: Can be wrapped with [FormControl (Select example)](https://chakra-ui.com/docs/components/form-control/usage#select-example)

File: apps/dashboard/pages/[endpoint].tsx [hunk 5]
```diff
@@ -236,6 +288,110 @@ const ModelPage: FC<ModelPageProps> = ({ endpoint }) => {
...
+                        })}
+                      </Select>
+                    </Box>
+                  </Flex>
+
+                  <Button
+                    disabled={isLoading}
+                    type='submit'
+                  >
+                    Generate
```
Comment (7alip) [inline apps/dashboard/pages/[endpoint].tsx:357]: You can add translations to all `common.json` files in each project.

```suggestion
                   {t('generate')}
```

Comment (7alip) [reply apps/dashboard/pages/[endpoint].tsx:357]: The simplest way to add a new translation to the common.json files can be to use VSCode's global search and replace tool.

Search

```jsonc
"give-feedback": "Give feedback",
```

```jsonc
// "give-feedback": "Feedback geven",
// "give-feedback": "Geri Bildirimde Bulun",
```

Replace

```json
"generate": "Generate",
"give-feedback": "Give feedback",
```

File: apps/dashboard/pages/api/route-tweet-gen.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+export const runtime = 'edge';
+ 
+const openai = new OpenAI({
+  apiKey: process.env.OPENAI_API_KEY!,
+});
+ 
+export default async function handler(req: Request) {
+  // Extract the `prompt` from the body of the request
+  const { prompt, numberOfPosts, charLimit, language } = await req.json();
+
```
Comment (7alip) [inline apps/dashboard/pages/api/route-tweet-gen.ts:13]: It seems some of the files are not formatted and there are whitespaces in the empty lines. You can investigate if it's default behavior of your editor.

If `format on save` setting doesn't work in your editor, you can run `yarn format` to format all the files in the project before you push.

```suggestion

```

File: apps/dashboard/pages/tweet-gen.tsx [hunk 7]
```diff
@@ -0,0 +1,63 @@
+'use client';
```
Comment (7alip) [inline apps/dashboard/pages/tweet-gen.tsx:1]: This test page could be removed

File: apps/samenvvv/pages/tweet-gen.tsx [hunk 8]
```diff
@@ -0,0 +1,63 @@
+'use client';
```
Comment (7alip) [inline apps/samenvvv/pages/tweet-gen.tsx:1]: This test page could be removed

File: .vscode/settings.json [hunk 9]
```diff
@@ -3,5 +3,6 @@
     "source.fixAll.eslint": "explicit"
   },
   "editor.formatOnSave": true,
-  "cSpell.words": ["strapi"]
+  "cSpell.words": ["strapi"],
+  "liveServer.settings.port": 5501
```
Comment (7alip) [inline .vscode/settings.json:7]: You can stash these changes if you will need them in the future.

File: apps/api/config/database.ts [hunk 10]
```diff
@@ -8,7 +8,7 @@ export default ({ env }) => {
       client: 'postgres',
       connection: {
         host: env('DATABASE_HOST', host),
-        port: '5432',
+        port: '5433',
```
Comment (7alip) [inline apps/api/config/database.ts:11]: Please revert this and stash if you need

File: apps/api/docker-compose.yml [hunk 11]
```diff
@@ -49,7 +49,7 @@ services:
       # copy the sql script to create tables
       - ./sql/init.sql:/docker-entrypoint-initdb.d/init-1.sql
     ports:
-      - '5432:5432'
+      - '5433:5433'
```
Comment (7alip) [inline apps/api/docker-compose.yml:52]: Should be reverted

File: apps/dashboard/pages/[endpoint].tsx [hunk 12]
```diff
@@ -50,6 +69,37 @@ const ModelPage: FC<ModelPageProps> = ({ endpoint }) => {
...
+    body: {
+      numberOfPosts,
+      charLimit,
+      language,
+    },
+    onFinish(prompt, completion) {
+      setGeneratedPosts(JSON.parse(completion))
+    },
+    onError(error) {
+      console.log("Error: ", error.message.substring(0, 280));
```
Comment (7alip) [inline apps/dashboard/pages/[endpoint].tsx:97]: You can use `toast` here for the users. Just display a generic error message `t('contact.form.failed')`. _(We can change the translation key to a generic key later.)_

File: apps/samenvvv/pages/api/route-tweet-gen.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+  const { prompt, numberOfPosts, charLimit } = await req.json();
+ 
+  // Request the OpenAI API for the response based on the prompt
+  const response = await openai.chat.completions.create({
+    model: 'gpt-4',
+    stream: true,
+    messages: [
+      {
+        role: 'system',
+        content: 'You are an activist, and your task is to raise awareness about human rights violations.'
```
Comment (7alip) [inline apps/samenvvv/pages/api/route-tweet-gen.ts:21]: Would it generate relative responses if alternative jobs were used, such as journalist, author, poet, and so on? If so, we can make it dynamic. It doesn't have to be in the UI now.

```ts
const { role = 'activist' } = await req.json()

content: `You are an ${role}, ...`
```

**Review by 7alip — CHANGES_REQUESTED** [review]
Excellent job, thanks a lot.

File: apps/dashboard/pages/[endpoint].tsx [hunk 14]
```diff
@@ -236,6 +288,110 @@ const ModelPage: FC<ModelPageProps> = ({ endpoint }) => {
         >
           {endpoint === 'posts' && selectedModel && (
             <Box p={4} rounded="md" bg="white" shadow="md">
+              {/* todo: button to generate posts using AI */}
+              <Heading p={4} color='black'>Post Generator</Heading>
+              <Box p={4}>
```
Comment (7alip) [inline apps/dashboard/pages/[endpoint].tsx:293]: Could you please move these sections to a separate component, as we did with `PostSentenceForm`? You should probably also move the `useCompletion` hook part and all associated states to the new component. We can keep the `[endpoint]` page clean using this method.

This component may be useful elsewhere in the project, such as in samen.

File: apps/api/package.json [hunk 15]
```diff
@@ -4,12 +4,7 @@
   "version": "0.1.0",
   "description": "A Strapi application",
   "scripts": {
```
Comment (7alip) [inline apps/api/package.json:6]:
```
    "dev": "strapi dev",
    "dashboard": "strapi dev",
    "foundation": "strapi dev",
    "kunsthalte": "strapi dev",
    "samenvvv": "strapi dev",
```   

We need these script

File: apps/foundation/public/locales/en/common.json [hunk 16]
```diff
@@ -131,6 +131,7 @@
   "foundation.assets": "Assets",
   "fullname": "Your Full Name",
   "give-feedback": "Give feedback",
+"generate": "Generate",,
```
Comment (7alip) [inline apps/foundation/public/locales/en/common.json:134]:
```suggestion
"generate": "Generate",
```

File: apps/kunsthalte/public/locales/en/common.json [hunk 17]
```diff
@@ -131,6 +131,7 @@
   "foundation.assets": "Assets",
   "fullname": "Your Full Name",
   "give-feedback": "Give feedback",
+"generate": "Generate",,
```
Comment (7alip) [inline apps/kunsthalte/public/locales/en/common.json:134]:
```suggestion
"generate": "Generate",
```

File: apps/lotus/public/locales/en/common.json [hunk 18]
```diff
@@ -131,6 +131,7 @@
   "foundation.assets": "Assets",
   "fullname": "Your Full Name",
   "give-feedback": "Give feedback",
+"generate": "Generate",,
```
Comment (7alip) [inline apps/lotus/public/locales/en/common.json:134]:
```suggestion
"generate": "Generate",
```

File: apps/samenvvv/public/locales/en/common.json [hunk 19]
```diff
@@ -131,6 +131,7 @@
   "foundation.assets": "Assets",
   "fullname": "Your Full Name",
   "give-feedback": "Give feedback",
+"generate": "Generate",,
```
Comment (7alip) [inline apps/samenvvv/public/locales/en/common.json:134]:
```suggestion
"generate": "Generate",
```

File: packages/ui/src/post-maker/TweetGenAI/TweetGenAI.tsx [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+          <Button type="button" onClick={stop} ml={2} colorScheme="gray">
+            Stop
+          </Button>
+        </form>
+      </Box>
+      {isLoading ? (
+        <Box p={4}>
+          <Progress size="xs" isIndeterminate />
+          {/* <div>{completion}</div> */}
+        </Box>
```
Comment (7alip) [inline packages/ui/src/post-maker/TweetGenAI/TweetGenAI.tsx:135]: Shall we display the completion for an effect? :)

```suggestion
        <Stack p={4}>
          <Progress size="xs" isIndeterminate />
          <Box color={'gray.500'}>{completion}</Box>
        </Box>
```

**Review by 7alip — CHANGES_REQUESTED** [review]

File: packages/ui/src/post-maker/TweetGenAI/TweetGenAI.tsx (hunk 20 above)
Comment (hakanayata) [reply packages/ui/src/post-maker/TweetGenAI/TweetGenAI.tsx:135]: I'll replace completion for the progress bar then. 😎

**Review by 7alip — APPROVED** [review]
Thank you!

File: apps/dashboard/pages/[endpoint].tsx [hunk 21]
```diff
@@ -236,6 +237,7 @@ const ModelPage: FC<ModelPageProps> = ({ endpoint }) => {
         >
           {endpoint === 'posts' && selectedModel && (
             <Box p={4} rounded="md" bg="white" shadow="md">
+              <TweetGenAI />
```
Comment (7alip) [inline apps/dashboard/pages/[endpoint].tsx:240]:
```suggestion
              <TweetGenAI content={(selectedContent as Post).content} />
```

If you want, you can still pass post's content data to use the default content.

File: packages/ui/src/post-maker/TweetGenAI/TweetGenAI.tsx [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+  const {
+    completion,
+    input,
+    stop,
+    isLoading,
+    handleInputChange,
+    handleSubmit,
+  } = useCompletion({
+    api: 'api/route-tweet-gen',
+    body: {
```
Comment (7alip) [inline packages/ui/src/post-maker/TweetGenAI/TweetGenAI.tsx:41]:
```suggestion
    initialInput: content,
    body: {
```

File: packages/ui/src/post-maker/TweetGenAI/TweetGenAI.tsx [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+      numberOfPosts,
+      charLimit,
+      language,
+    },
+    onFinish(prompt, completion) {
+      setGeneratedPosts(JSON.parse(completion))
+    },
+    onError(error) {
+      toastMessage('Error', t('contact.form.failed'), 'error')
+    },
```
Comment (7alip) [inline packages/ui/src/post-maker/TweetGenAI/TweetGenAI.tsx:51]:
```suggestion
    onFinish(prompt: string, completion: string) {
      setGeneratedPosts(JSON.parse(completion))
    },
    onError() {
      toastMessage('Error', t('contact.form.failed'), 'error')
    },
```

File: packages/ui/src/post-maker/TweetGenAI/TweetGenAI.tsx [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+        <form onSubmit={handleSubmit}>
+          <FormControl>
+            <FormLabel>Content</FormLabel>
+            <Textarea
+              name="prompt"
+              placeholder="Enter a content..."
+              value={input}
+              onChange={handleInputChange}
+              mb={4}
+              size="xs"
```
Comment (7alip) [inline packages/ui/src/post-maker/TweetGenAI/TweetGenAI.tsx:69]: You can remove `size`, it looks so small.

File: apps/samenvvv/pages/api/route-tweet-gen.ts (hunk 13 above)
Comment (hakanayata) [reply apps/samenvvv/pages/api/route-tweet-gen.ts]: It seems like the profession plays a negligible role. The sentence after the comma does matter the most.

Comment (7alip) [reply apps/samenvvv/pages/api/route-tweet-gen.ts]: Thanks, I'll think about this later.

**Pull request merged**